File: Builder/Exercicios/Ex3_Roupa/Ex3.py
"""
PADRÃO: BUILDER

Exercício: Fábrica de Roupas customizadas.
O Builder permite encapsular a lógica de criação de objetos complexos 
(Roupa), permitindo que o cliente defina os atributos de forma fluída.
"""

# 1. COMPONENTES DO PRODUTO
class Tipo:
    def __init__(self, tipo): self.tipo = tipo

class Tamanho:
    def __init__(self, tamanho): self.tamanho = tamanho

class Cor:
    def __init__(self, cor): self.cor = cor

class Preco:
    def __init__(self, preco): self.preco = preco

# 2. PRODUTO FINAL
class Roupa:
    def __init__(self, tipo, cor, tamanho, preco):
        self.tipo = tipo
        self.cor = cor
        self.tamanho = tamanho
        self.preco = preco

    def __str__(self):
        return f"Roupa [Cor: {self.cor.cor}, Tipo: {self.tipo.tipo}, Tamanho: {self.tamanho.tamanho}, Preço: R$ {self.preco.preco:.2f}]"

# 3. BUILDER CONCRETO
class RoupaBuilder:
    def __init__(self):
        self.tipo = None
        self.cor = None
        self.tamanho = None
        self.preco = None

    def add_tipo(self, tipo):
        self.tipo = Tipo(tipo)
        return self

    def add_cor(self, cor):
        self.cor = Cor(cor)
        return self

    def add_tamanho(self, tamanho):
        self.tamanho = Tamanho(tamanho)
        return self

    def add_preco(self, preco):
        self.preco = Preco(preco)
        return self

    def construir(self):
        return Roupa(self.tipo, self.cor, self.tamanho, self.preco)

# 4. CLIENTE
if __name__ == "__main__":
    builder = RoupaBuilder()
    
    # Criando uma Calça Jeans
    roupa1 = (builder
              .add_cor("Azul")
              .add_preco(89.90)
              .add_tamanho("42")
              .add_tipo("Calça Jeans")
              .construir())

    print(roupa1)

File: Builder/Exercicios/Ex3_Roupa/Ex3.java
/**
 * PADRÃO: BUILDER
 * 
 * Exercício: Fábrica de Roupas customizadas.
 * O Builder permite definir cor, tipo, tamanho e preço de forma encadeada, 
 * tornando o processo de criação de objetos de Produto muito mais intuitivo.
 */

// 1. PARTES DO PRODUTO (Atributos da Roupa)
class Tipo {
    public String tipo;
    Tipo(String tipo) { this.tipo = tipo; }
}

class Tamanho {
    public String tamanho;
    Tamanho(String tamanho) { this.tamanho = tamanho; }
}

class Cor {
    public String cor;
    Cor(String cor) { this.cor = cor; }
}

class Preco {
    public float preco;
    Preco(float preco) { this.preco = preco; }
}

// 2. PRODUTO FINAL
class Roupa {
    public Preco preco;
    public Tamanho tamanho;
    public Cor cor;
    public Tipo tipo;

    Roupa(Tipo tipo, Cor cor, Tamanho tamanho, Preco preco) {
        this.cor = cor;
        this.tipo = tipo;
        this.tamanho = tamanho;
        this.preco = preco;
    }

    @Override
    public String toString() {
        return String.format("Cor: %s | Tipo: %s | Tamanho: %s | Preço: R$ %.2f", 
                             this.cor.cor, this.tipo.tipo, this.tamanho.tamanho, this.preco.preco);
    }
}

// 3. BUILDER CONCRETO
class RoupaBuilder {
    private Preco preco;
    private Tamanho tamanho;
    private Cor cor;
    private Tipo tipo;

    public RoupaBuilder addPreco(float preco) {
        this.preco = new Preco(preco);
        return this;
    }

    public RoupaBuilder addTamanho(String tamanho) {
        this.tamanho = new Tamanho(tamanho);
        return this;
    }

    public RoupaBuilder addCor(String cor) {
        this.cor = new Cor(cor);
        return this;
    }

    public RoupaBuilder addTipo(String tipo) {
        this.tipo = new Tipo(tipo);
        return this;
    }

    public Roupa construir() {
        return new Roupa(this.tipo, this.cor, this.tamanho, this.preco);
    }
}

// 4. CLIENTE
public class Ex3 {
    public static void main(String[] args) {
        RoupaBuilder builder = new RoupaBuilder();

        // Criando uma Camiseta Roxa
        Roupa roupa1 = builder
            .addCor("Roxo")
            .addPreco(29.90f)
            .addTamanho("M")
            .addTipo("Camiseta")
            .construir();

        System.out.println(roupa1);
    }
}
